providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/resolver/process/storage/connector/sync/SyncStreamQueueSource.java

```diff
@@ -0,0 +1,174 @@
+
+        log.info("Initializing sync stream observer");
+
+        // outer loop for re-issuing the stream request
+        while (!shutdown.get()) {
+
+            log.debug("Initializing sync stream request");
+            final GetMetadataRequest.Builder metadataRequest = GetMetadataRequest.newBuilder();
+            GetMetadataResponse metadataResponse = GetMetadataResponse.getDefaultInstance();
+            restart();
+
+            try (CancellableContext context = Context.current().withCancellation()) {
+
+                try {
+                    metadataResponse = channelConnector.getBlockingStub().getMetadata(metadataRequest.build());
+                } catch (Exception metaEx) {
+                    log.error("Metadata exception: {}", metaEx.getMessage(), metaEx);
+                    context.cancel(metaEx);
+                }
+
+                // inner loop for handling messages
+                while (!shutdown.get()) {
+                    final SyncResponseModel taken = incomingQueue.take();
+                    if (taken.isComplete()) {
+                        log.debug("Sync stream completed, will reconnect");
+                        // The stream is complete, we still try to reconnect
+                        break;
+                    }
+
+                    Throwable streamException = taken.getError();
+                    if (streamException != null) {
+                        log.debug("Exception in GRPC connection, streamException {}, will reconnect", streamException);
+                        if (!outgoingQueue.offer(new QueuePayload(
+                                QueuePayloadType.ERROR, "Error from stream or metadata", metadataResponse))) {
+                            log.error("Failed to convey ERROR status, queue is full");
+                        }
+                        break;
+                    }
+
+                    final SyncFlagsResponse flagsResponse = taken.getSyncFlagsResponse();
+                    final String data = flagsResponse.getFlagConfiguration();
+                    log.debug("Got stream response: {}", data);
+
+                    if (!outgoingQueue.offer(new QueuePayload(QueuePayloadType.DATA, data, metadataResponse))) {
+                        log.error("Stream writing failed");
+                    }
+                }
+            }
+        }
+
+        log.info("Shutdown invoked, exiting event stream listener");
+    }
+
+    private void restart() {
+        FlagSyncServiceStub localStub = stub; // don't mutate the stub
+        if (streamDeadline > 0) {
+            localStub = localStub.withDeadlineAfter(streamDeadline, TimeUnit.MILLISECONDS);
+        }
+
+        final SyncFlagsRequest.Builder syncRequest = SyncFlagsRequest.newBuilder();
+        if (this.selector != null) {
+            syncRequest.setSelector(this.selector);
+        }
+
+        if (this.providerId != null) {
+            syncRequest.setProviderId(this.providerId);
```

Comment:
@cupofcat FYI some of your recent changes moved here. I didn't change your tests and I'm confident everything still works, I just didn't want you to be confused if you noticed your changes were immediate moved :sweat_smile: